### ISSUE TITLE
🚀 Use something like Graphtage for specific file types

### ISSUE THREAD

**Issue by boris-petrov** [dandavison/delta#535]
I stumbled upon [Graphtage](https://github.com/trailofbits/graphtage) and found it a cool idea. It would be nice if `delta` could integrate it (and other projects like it) for specific file types. What do you think?

**Comment by dandavison** [comment]
Hi @boris-petrov sorry to be so slow replying to this one.

Can you be explicit about what sort of integration might make sense? It doesn't look like graphtage produces diff format AFAICS. So are you thinking of any of the following?

1. User can do `graphtage a.json b.json | delta` (but how is that any better than `graphtage a.json b.json`?
2. User can do `delta a.json b.json` and have graphtage-like output (ok, but that is tantamount to re-implementing graphtage in Rust; delta is really a diff *viewer* not a diff *generator*)
3. Something else?

**Comment by bew** [comment]
FYI there's a project similar to delta that does exactly this (in Rust):
https://github.com/Wilfred/difftastic
(not as pretty/configurable as delta though)

Maybe there's some inspiration to take here?

**Comment by dandavison** [comment]
> FYI there's a project similar to delta that does exactly this (in Rust): https://github.com/Wilfred/difftastic

Right, difftastic looks great. I'd seen it before and meant to look more closely. One form of integration there could perhaps be difftastic emitting JSON and delta displaying it.

**Comment by marcelarie** [comment]
Is the support for difftastic available?

**Comment by marcelarie** [comment]
Ok thanks for the info. As far as I know, JSON is not the fastest to parse, right? Maybe something more raw would be better 🤷🏽

**Comment by Guekka** [comment]
> As far as I know, JSON is not the fastest to parse, right? Maybe something more raw would be better. 🤷🏽

I don't think that's an issue. Parsing JSON [can be very fast](https://github.com/simdjson/simdjson). And diffing is way more expensive than any parsing could be

**Comment by marcelarie** [comment]
The PR on `difftastic` to output JSON has been successfully merged: https://github.com/Wilfred/difftastic/pull/158#issuecomment-1682537893

**Comment by ding-weibing** [comment]
> > Is the support for difftastic available?
> 
> Not as far as I know. Looks like one path would be if a PR like this one is merged in difftastic: [Wilfred/difftastic#158](https://github.com/Wilfred/difftastic/pull/158)
> 
> and then someone implements rendering of that JSON in delta. Delta already renders `ripgrep --json` output, so perhaps that implementation will be helpful when working on the difftastic JSON. (I guess in an ideal world there would be a standard format for describing token-level diffs regardless of their provenance; perhaps then difftastic and `git diff --word-diff` output could be the same format??)

Hi @dandavison,

Based on the recent discussion at [Delta Issue #535](https://github.com/dandavison/delta/issues/535#issuecomment-1682577098), it seems there might be an opportunity to address the support issue with difftastic. While I am not highly familiar with Rust, I am eager to contribute to this effort. If you could guide me on which parts of the code require modifications, I would be delighted to make a commit. Thank you for your assistance and insight.

**Comment by Kurren123** [comment]
This would be great to support as difftastic allows reading diffs with much less mental overhead. For me in C# it eliminates so much noise.

**Comment by Explosion-Scratch** [comment]
I would love this feature and seeing as difftastic now supports json I don't see why it wouldn't be possible

**Comment by serramatutu** [comment]
Hey all! 

I have been working on implementing this as a generic `ExternalDiff` state that can be used to parse the output from any external diff program into what `delta` uses to render (via `git diff --ext-diff` or something like `difft --display json a.rs b.rs | delta`.

For difftastic speficially, I stumbled upon an issue: the difftastic JSON output only shows changes, but does not contain the surrounding context for the change. This means you cannot pull the context out of difftastic's output by itself (or `git diff --ext-diff`). In contrast, `ripgrep`'s JSON output _does_ include the surrounding context, and this makes the implementation on `delta`'s side much simpler.

There are 2 options:
1. Make `delta` pull the context of the hunk around the change. I don't like this as it means delta will have to re-open and re-analyze the contents of each file that got changed and a lot of the logic will be reimplemented again
2. Make `difftastic` provide the context around the change. Unless we create a new `--display json-hunk` mode, this will be a breaking change as currently difftastic's output is line-based and not hunk-based, i.e it produces multiple entries for _each_ change, while `delta` is interested in _one_ single hunk with stuff tagged in it. Maybe a breaking change is fine tho as JSON output is marked as experimental anyways.

If we choose option 1, we need to make both the base and compare filenames available in the output of diffstatic so that delta can load them. Currently, only the compare filename is available (see output at the end of my comment).

If we choose option 2 (IMO the better one), we first have to merge the changes to make the context available into difftastic and only then work on delta. 

Either way, it will depend on the maintainer of difftastic agreeing with changes. I will try to submit Option 2 to the difftastic repo (once I'm done with it), and see what they think about it. Meanwhile, I'm afraid the progress on `delta` is blocked.

However, if anyone is interested, I can try to push my `ExternalDiff` changes across the line (without difftastic support) so people can implement Graphtage or other programs on top of that :)

I wanted to hear your thoughts!

### What I got so far

1. Begin implementing external diff (starting with difftastic) support in delta: https://github.com/serramatutu/delta/tree/serramatutu/difftastic
  - TODO: implement `handle_external_diff` properly, implement state machine transitions
2. Change difftastic's JSON output to use hunks instead of lines: https://github.com/serramatutu/difftastic/tree/serramatutu/src-line-json
  - TODO: remove the inner for loop in `json.rs` and replace the structs to make it spit out hunks instead of lines.
  - QUESTION: should this really be a breaking change? Should we add `--display json-hunks` instead? Should we make `delta` re-read the files?

### Example

Let's say I changed a file from:
```rust
// comment

pub fn main() {
    let a = 123u8;
    func(a + 1);
}
```
 
to:
```rust
// comment

pub fn main() {
    let a = 123u8;
    func(a  - 1);
}
```

What `delta` would need to know:
- we replaced `+` with `-` (and where the replace starts/ends to know where to highlight)
- the context to display around it (in this case, 2 lines before and 1 after)
- the file names before and after (for file renames)
- sufficient information to calculate line numbers to display

It would need to be something like:
```json
{
  "hunks": [
    {
      "status": "changed",
      "lhs": {
        "start_line": 2,
        "num_lines": 4,
        "src": "pub fn main() {\n    let a = 123u8;\n    func(a + 1);\n}",
        "path": "b.rs"
      },
      "rhs": {
        "start_line": 2,
        "num_lines": 4,
        "src": "pub fn main() {\n    let a = 123u8;\n    func(a - 1);\n}",
        "path": "b.rs"
      },
      "changes": [
        {
          "lhs" : {
             "content": "+",
             "start": 46,
             "end": 47
          },
          "rhs" : {
             "content": "-",
             "start": 46,
             "end": 47
          }
        }
      ]
    }
  ]
}
```

This is what `difftastic` currently outputs. You can see the information is insufficient.
```json
{
  "chunks": [
    [
      {
        "lhs": {
          "line_number": 4,
          "changes": [
            {
              "start": 11,
              "end": 12,
              "content": "+",
              "highlight": "normal"
            }
          ]
        },
        "rhs": {
          "line_number": 4,
          "changes": [
            {
              "start": 11,
              "end": 12,
              "content": "-",
              "highlight": "normal"
            }
          ]
        }
      }
    ]
  ],
  "language": "Rust",
  "path": "b.rs",
  "status": "changed"
}
```

**Comment by lucaspar** [comment]
(Outsider perspective, so I'm ready to be wrong, but...)

Isn't start_line and num_lines redundant with the source code?

Surely we can retrieve the source from that information, which would also allow dynamically changing the context window without rerunning the diff.

**Comment by serramatutu** [comment]
AFAIK delta can't change the context window dynamically. It fetches the context from the diff program (ie git diff), or in the case of ripgrep it comes from ripgrep itself.

If you run `rg -B 3 -A 3 --json`, the context window will be -3+3, coming from ripgrep itself. Delta doesn't re-open the file to make its own context window.

**Comment by farnoy** [comment]
I think delta also runs its own rename detection when it's operating as a pager? It would help with a long-standing limitation in jujutsu -  https://github.com/jj-vcs/jj/issues/47. With delta calling out to an external diff tool, you could use any one you want without sacrificing rename detection.

**Comment by dandavison** [comment]
Nice work @serramatutu!

> I think delta also runs its own rename detection when it's operating as a pager?

The basic design of delta is that it is a Unix "[filter](https://en.wikipedia.org/wiki/Filter_(software))": it receives diff/ripgrep output on stdin, and emits decorated output on stdout. Its stdout is sent to `less`. So it would be nicest if difftastic could pass this context on. (Have difftastic users not asked to see context around changes?)


Delta has some documentation of its internals: https://github.com/dandavison/delta/blob/main/ARCHITECTURE.md

**Comment by farnoy** [comment]
Here's what I observed:

```sh
$ jj config list ui
ui.diff-formatter = "difft"
ui.pager = "less -FRX"
$ jj diff -r $revset --tool delta | grep -i renamed | wc -l
52
$ jj diff -r $revset --tool difft | grep -i renamed | wc -l
0
# jj rename detection is incomplete, hence:
$ jj diff --git -r $revset | grep -i "rename from" | wc -l
21
$ jj diff --git -r $revset | delta | grep -i renamed | wc -l
21
$ git config get core.pager
/nix/store/nzqhbhaxfxaywsx9x3pach6hbqpxhvgb-delta-0.18.2/bin/delta
$ git diff $revset | delta | grep -i renamed | wc -l
52
$ git -c diff.external=difft diff $revset | grep -i renamed | wc -l
52
$ git -c core.pager="less -FRX" $revset | grep -i "rename from" | wc -l
52
```

I only tried the first two `jj diff` commands before my previous comment, which is why I thought it was because delta receives the full input and does something extra on top, covering for jj. But now I don't know anymore

**Comment by abcdw** [comment]
@serramatutu From what I understand, a new `--display` format for difftastic should be the best way to proceed here.

Cc: @Wilfred

**Comment by asukaminato0721** [comment]
my daily usage is use delta as my default lazygit diff.

and bind a "D" to use difft for those a LOT of ++ and --

since difft don't have proper highlight (only diff highlight)

**Comment by serramatutu** [comment]
Hey everyone! I know I said I would work on this but life's been busy and I'm not annoyed enough by my terminal colors to go implement all of this :P

I probably won't keep working on those branches I linked but there is a PoC there. Anyone who's interested, feel free to pick them up!

**Comment by dimanovikov** [comment]
For structured *data* (as opposed to source code), datadiff (https://github.com/dimanovikov/datadiff) might fit this niche: semantic diff for JSON/YAML/CSV/TOML/XML — ignores key order and formatting, matches arrays of objects by a key field (`--key id`), and reports changes as data paths. It deliberately avoids optimal tree-edit-distance algorithms (Graphtage's O(N⁴)) in favor of predictable structural matching, so a 5,000-object JSON diffs in ~0.1s. It can already act as a delta-adjacent external diff via `git config diff.datadiff.command` + .gitattributes, and has `--output json` if a tool wants to render the changes itself.

**Comment by WieeRd** [comment]
@cloudroad-io While it is visible in the URL, you should always make it clear that this is a self-promotion whenever you bring up your own project.
It is also worth noting that the project was heavily vibe coded.

**Comment by dimanovikov** [comment]
Fair point — I'm the author of datadiff and should have said so explicitly. I'll disclose it going forward. On how it's built: yes, it's AI-assisted, but it ships with a black-box integration test suite, CI across Linux/Windows/macOS, and the whole thing is small enough to review in one sitting. Happy to be judged on the code and tests rather than the process.
